packages/issuer-token-endpoint/lib/IssuerTokenEndpoint.ts

```diff
@@ -0,0 +1,134 @@
+  nonceStateManager?: IStateManager<CNonceState>
+  jwtSignerCallback?: JWTSignerCallback
+}
+
+let tokenPath: string
+let interval: number
+let cNonceExpiresIn: number
+let tokenExpiresIn: number
+let stateManager: IStateManager<CredentialOfferState>
+let nonceStateManager: IStateManager<CNonceState>
+let jwtSignerCallback: JWTSignerCallback
+
+export const tokenRequestEndpoint = (opts?: ITokenEndpointOpts): Router => {
+  tokenPath = (opts?.tokenPath ? opts.tokenPath : (process.env.TOKEN_PATH as string)) ?? '/token'
+  interval = (opts?.interval ? opts.interval : getNumberOrUndefined(process.env.INTERVAL)) ?? 300000
+  cNonceExpiresIn = (opts?.cNonceExpiresIn ? opts.cNonceExpiresIn : getNumberOrUndefined(process.env.C_NONCE_EXPIRES_IN)) ?? 300000
+  tokenExpiresIn = (opts?.tokenExpiresIn ? opts.tokenExpiresIn : getNumberOrUndefined(process.env.TOKEN_EXPIRES_IN)) ?? 300000
+  if (opts?.jwtSignerCallback) {
+    jwtSignerCallback = opts.jwtSignerCallback
+  } else {
+    throw new Error('Please provide a private a JWT signer function')
+  }
+  if (opts?.stateManager) {
+    stateManager = opts.stateManager
+  } else {
+    throw new Error('Unable to proceed without an StateManager instance')
+  }
+  if (opts.nonceStateManager) {
+    nonceStateManager = opts.nonceStateManager
+  } else {
+    throw new Error('Unable to proceed without an NonceStateManager instance')
+  }
+  router.post(tokenPath, handleHTTPStatus400, handleTokenRequest)
+  return router
+}
+
+const generateAccessToken = async (): Promise<string> => {
+  const issuanceTime = new Date()
+  const jwt: Jwt = {
+    header: { typ: Typ.JWT, alg: Alg.ES256 },
+    payload: { iat: issuanceTime.getTime(), exp: +new Date(issuanceTime.getTime() + tokenExpiresIn) },
+  }
+  return await jwtSignerCallback(jwt)
+}
+
+const handleTokenRequest = async (request: Request, response: Response) => {
+  response.set({
+    'Cache-Control': 'no-store',
+    Pragma: 'no-cache',
+  })
+
+  const cNonce = v4()
+  await nonceStateManager.setState(cNonce, { cNonce, createdOn: +new Date() })
+  setTimeout(() => {
+    nonceStateManager.deleteState(cNonce)
+  }, cNonceExpiresIn)
+
+  const access_token = await generateAccessToken()
+  const responseBody: AccessTokenResponse = {
+    access_token,
+    token_type: 'bearer',
+    expires_in: tokenExpiresIn,
+    c_nonce: cNonce,
+    c_nonce_expires_in: cNonceExpiresIn,
+    authorization_pending: false,
+    interval,
+  }
+  return response.status(200).json(responseBody)
+}
+
+export const handleHTTPStatus400 = async (request: Request, response: Response, next: NextFunction) => {
+  // TODO invalid_client: the client tried to send a Token Request with a Pre-Authorized Code without Client ID but the Authorization Server does not support anonymous access
+  const assertedState = (await stateManager.getAssertedState(request.body.state)) as CredentialOfferState
+  await stateManager.setState(request.body.state, assertedState)
```

Comment:
Wait what? Why are we updating the state which we just got with a value coming in from the client, which we do not trust to begin with. Doesn't make any sense to me

Reply:
Just garbage left behind